test/integration_test.py

```diff
@@ -735,34 +755,74 @@ def _assert_indices_exist(self, catalog: CatalogName):
         for index_name in service.index_names(catalog):
             self.assertTrue(es_client.indices.exists(index_name))
 
-    def _test_list_sources(self, catalog: CatalogName):
-
-        def _list_sources(http_client: RequestMethods):
-            response = http_client.request(
-                'GET',
-                furl(config.service_endpoint(),
-                     path='/repository/sources',
-                     query={'catalog': catalog}).url
-            )
-            return set(freeze(json.loads(response.data)['sources']))
+    def _list_sources(self, catalog: CatalogName) -> Set[frozendict]:
+        url = furl(config.service_endpoint(),
+                   path='/repository/sources',
+                   query={'catalog': catalog}).url
+        response = self._get_url_json(url)
+        sources = freeze(response['sources'])
+        assert isinstance(sources, tuple)
+        return set(sources)
 
+    def _test_managed_access(self, catalog: CatalogName):
         plugin = self.azul_client.repository_plugin(catalog)
         assert isinstance(plugin, tdr.Plugin)
         # Uses the indexer service account credentials, which should have access
         # to all sources.
         tdr_client = plugin.tdr
-        all_sources = {
+        configured_sources = {
             frozendict(sourceSpec=str(source_spec),
                        sourceId=tdr_client.lookup_source_id(source_spec))
             for source_spec in map(TDRSourceSpec.parse, config.tdr_sources(catalog))
         }
-        self.assertEqual(_list_sources(tdr_client._http_client), all_sources)
+        public_sources = self._list_sources(catalog)
+        with self._service_account_credentials():
```

Comment:
I think there is one superfluous indirection. This should read

```suggestion
        with self._service_account_credentials:
```

It seems odd to cache a callable. I could be wrong.

Reply:
Removing the parens results in
```
======================================================================
ERROR: test_indexing (integration_test.IndexingIntegrationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/nadove/PycharmProjects/azul-sc/test/azul_test_case.py", line 142, in wrapped
    return method(*args, **kwargs)
  File "/home/nadove/PycharmProjects/azul-sc/test/integration_test.py", line 268, in test_indexing
    self._test_managed_access(catalog.name)
  File "/home/nadove/PycharmProjects/azul-sc/test/integration_test.py", line 789, in _test_managed_access
    with self._service_account_credentials:
AttributeError: __enter__
```

`_service_account_credentials` is a property that returns a function that returns a context manager:

```
>>> from integration_test import IndexingIntegrationTest; i = IndexingIntegrationTest(); i.setUp()
>>> i._service_account_credentials
<function IndexingIntegrationTest._authorization_context.<locals>.func at 0x7ff0a15641f0>
>>> i._service_account_credentials()
<contextlib._GeneratorContextManager object at 0x7ff0a6e0f340>
```

The extra level of indirection allows us to create context managers for the service account and public service account credentials with minimal duplication. The caching prevents the local function in `_authorization_context` from being re-created every time the method is called:

```
>>> i._service_account_credentials()
<contextlib._GeneratorContextManager object at 0x7ff0a6e0f340>
>>> i._service_account_credentials()
<contextlib._GeneratorContextManager object at 0x7ff0a6e0f340>
```

as opposed to:

```
$ git diff
diff --git a/test/integration_test.py b/test/integration_test.py
index 5c8d94f6..8bb27ff5 100644
--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -367,11 +367,11 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
             self._test_dos(catalog, file_uuid)
             self._test_drs(catalog, file_uuid)
 
-    @cached_property
+    @property
     def _service_account_credentials(self):
         return self._authorization_context(TDRClient.with_service_account_credentials())
 
-    @cached_property
+    @property
     def _public_service_account_credentials(self):
         return self._authorization_context(TDRClient.with_public_service_account_credentials())
 
$ python3
Python 3.8.3 (default, May 22 2020, 15:24:35) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from integration_test import IndexingIntegrationTest; i = IndexingIntegrationTest(); i.setUp()
>>> i._service_account_credentials
<function IndexingIntegrationTest._authorization_context.<locals>.func at 0x7f4eeb0301f0>
>>> i._service_account_credentials
<function IndexingIntegrationTest._authorization_context.<locals>.func at 0x7f4eeb030310>
>>> i._service_account_credentials
<function IndexingIntegrationTest._authorization_context.<locals>.func at 0x7f4eeb030160>
```

Note that we don't and can't cache the context manager itself because it is not reusable. If it were, then we could call `func` once, cache the result, and _then_ ditch the parens at the call sites. As it stands this just doesn't work due to the design of `contextlib`.

Reply:
That may all be true but we have to balance the goal of efficiency against that of writing easy to understand and maintainable code. I'm sure you agree that the current approach is rather complicated and difficult to comprehend. Anything wrong with the approach below?

```diff
Index: test/integration_test.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/integration_test.py b/test/integration_test.py
--- a/test/integration_test.py	(revision 1ddd6460b2f0637527406aea3893c6817c782594)
+++ b/test/integration_test.py	(date 1631208504712)
@@ -26,7 +26,6 @@
 from typing import (
     AbstractSet,
     Any,
-    Callable,
     ContextManager,
     Dict,
     IO,
@@ -202,12 +201,11 @@
         """
         Test with a small page size to be sure paging works
         """
-        tdr_client = TDRClient.with_public_service_account_credentials()
         page_size = 5
         with mock.patch.object(TDRClient, 'page_size', page_size):
-            paged_snapshots = tdr_client.snapshot_names_by_id()
+            paged_snapshots = self._public_tdr_client.snapshot_names_by_id()
         self.assertGreater(len(paged_snapshots), page_size)
-        snapshots = tdr_client.snapshot_names_by_id()
+        snapshots = self._public_tdr_client.snapshot_names_by_id()
         self.assertEqual(snapshots, paged_snapshots)
 
     def test_indexing(self):
@@ -273,7 +271,7 @@
                                        notifications=catalog.notifications_with_duplicates(),
                                        delete=True)
             _wait_for_indexer()
-            with self._service_account_credentials():
+            with self._service_account_credentials:
                 for catalog in catalogs:
                     self._assert_catalog_empty(catalog.name)
 
@@ -368,26 +366,30 @@
             self._test_drs(catalog, file_uuid)
 
     @cached_property
-    def _service_account_credentials(self):
-        return self._authorization_context(TDRClient.with_service_account_credentials())
+    def _tdr_client(self) -> TDRClient:
+        return TDRClient.with_service_account_credentials()
 
     @cached_property
-    def _public_service_account_credentials(self):
-        return self._authorization_context(TDRClient.with_public_service_account_credentials())
+    def _public_tdr_client(self) -> TDRClient:
+        return TDRClient.with_public_service_account_credentials()
 
-    def _authorization_context(self, tdr: TDRClient) -> Callable[[], ContextManager]:
+    @property
+    def _service_account_credentials(self) -> ContextManager:
+        return self._authorization_context(self._tdr_client)
+
+    @property
+    def _public_service_account_credentials(self) -> ContextManager:
+        return self._authorization_context(self._public_tdr_client)
 
-        @contextmanager
-        def func():
-            old_http = self._http
-            try:
-                self._http = tdr._http_client
-                yield
-            finally:
-                self._http = old_http
+    @contextmanager
+    def _authorization_context(self, tdr: TDRClient) -> ContextManager:
+        old_http = self._http
+        try:
+            self._http = tdr._http_client
+            yield
+        finally:
+            self._http = old_http
 
-        return func
-
     def _check_endpoint(self,
                         endpoint: str,
                         path: str,
@@ -697,7 +699,7 @@
             retries = 0
             deadline = time.time() + timeout
             while True:
-                with self._service_account_credentials():
+                with self._service_account_credentials:
                     hits = self._get_entities(catalog, entity_type)
                 indexed_fqids.update(
                     # FIXME: We should use the source from the index rather than
@@ -786,9 +788,9 @@
             for source_spec in map(TDRSourceSpec.parse, config.tdr_sources(catalog))
         }
         public_sources = self._list_sources(catalog)
-        with self._service_account_credentials():
+        with self._service_account_credentials:
             self.assertEqual(configured_sources, self._list_sources(catalog))
-        with self._public_service_account_credentials():
+        with self._public_service_account_credentials:
             self.assertEqual(public_sources, self._list_sources(catalog))
         self.assertEqual(public_sources - configured_sources, set())
 
@@ -815,7 +817,7 @@
         self.assertEqual(response['hits'], [])
         self.assertEqual(response['pagination']['total'], 0)
 
-        with self._service_account_credentials():
+        with self._service_account_credentials:
             response = self._get_url_json(url)
         hit_source_ids = _source_ids_from_hits(response['hits'])
         self.assertEqual(hit_source_ids, managed_access_source_ids)
@@ -828,7 +830,7 @@
             file_url = first(managed_access_files)
             response = self._get_url_unchecked(file_url, redirect=False)
             self.assertEqual(response.status, 404)
-            with self._service_account_credentials():
+            with self._service_account_credentials:
                 response = self._get_url_unchecked(file_url, redirect=False)
                 self.assertIn(response.status, (301, 302))
         else:
```